l1-contracts/src/governance/Gerousia.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.27;
+
+import {IRegistry} from "@aztec/governance/interfaces/IRegistry.sol";
+import {IApella} from "@aztec/governance/interfaces/IApella.sol";
+import {IGerousia} from "@aztec/governance/interfaces/IGerousia.sol";
+import {Errors} from "@aztec/governance/libraries/Errors.sol";
+
+import {Slot, SlotLib} from "@aztec/core/libraries/TimeMath.sol";
+import {ILeonidas} from "@aztec/core/interfaces/ILeonidas.sol";
+
+/**
+ * @notice  A Gerousia implementation following the empire model
+ *          Beware that while governance generally do not care about the implementation
+ *          this implementation will since it is dependent on the sequencer selection.
+ *          This also means that the implementation here will need to be "updated" if
+ *          the interfaces of the sequencer selection changes, for exampel going optimistic.
+ */
+contract Gerousia is IGerousia {
+  using SlotLib for Slot;
+
+  struct RoundAccounting {
+    Slot lastVote;
+    address leader;
+    bool executed;
+    mapping(address proposal => uint256 count) yeaCount;
+  }
+
+  uint256 public constant LIFETIME_IN_ROUNDS = 5;
+
+  IApella public immutable APELLA;
+  IRegistry public immutable REGISTRY;
+  uint256 public immutable N;
+  uint256 public immutable M;
+
+  mapping(address instance => mapping(uint256 roundNumber => RoundAccounting)) public rounds;
+
+  constructor(IApella _apella, IRegistry _registry, uint256 _n, uint256 _m) {
+    APELLA = _apella;
+    REGISTRY = _registry;
+    N = _n;
+    M = _m;
+
+    require(N > M / 2, Errors.Gerousia__InvalidNAndMValues(N, M));
+    require(N <= M, Errors.Gerousia__NCannotBeLargerTHanM(N, M));
+  }
+
+  // Note that this one is heavily realying on the fact that this contract
+  // could be updated at the same time as another upgrade is made.
+
+  /**
+   * @notice	Cast a vote on a proposal
+   *					Note that this is assuming that the canonical rollup will cast it as
+   * 					part of block production, we will perform it here
+   *
+   * @param _proposal - The proposal to cast a vote on
+   *
+   * @return True if executed successfully, false otherwise
+   */
+  function vote(address _proposal) external override(IGerousia) returns (bool) {
+    require(_proposal.code.length > 0, Errors.Gerousia__ProposalHaveNoCode(_proposal));
+
+    address instance = REGISTRY.getRollup();
+    require(instance.code.length > 0, Errors.Gerousia__InstanceHaveNoCode(instance));
+
+    ILeonidas selection = ILeonidas(instance);
+    Slot currentSlot = selection.getCurrentSlot();
+
+    uint256 roundNumber = computeRound(currentSlot);
+
+    RoundAccounting storage round = rounds[instance][roundNumber];
+
+    require(currentSlot > round.lastVote, Errors.Gerousia__VoteAlreadyCastForSlot(currentSlot));
+
+    address proposer = selection.getCurrentProposer();
+    require(msg.sender == proposer, Errors.Gerousia__OnlyProposerCanVote(msg.sender, proposer));
```

Comment:
Must we constrain proposers to vote during their turn?

Reply:
The Empire model is using the sampling to figure out who should cast votes to make a proposal.

So practically, you don't NEED, to do it as part of the proposer slot, but it have its benefits

- You don't have to handle some extra accounting for what slots have been "voted on" after the fact (this is not super expensive, can be done quite efficiently with some bitpacking). 
- More importantly, it means that the full set of "voters" are not yet know early, so it should be harder for you to coordinate a malicious pushthrough.
- By only needing the data for "now" we avoid some of the cases that could happen with new instances being deployed.

For example, say that as long as it is the proposers, they are allowed to vote later. You could wait until you see a round where you have sufficient "ballots" and then cast them all at once to make your proposal. In the current model, you would need to "publish" what your intent is early, and then hope that you get sufficient.

I don't particular care. In any case they are unable to force something through because that is handled fully separate to this sampling in the Apella.

The implementation was essentially made like this because that is what empire is proposing, and that was what you guys wanted 🤷

Reply:
Okay this means that for a proposal to be nominated, N / M validators must vote to propose it. This quorum is tied to the round size and no longer to the size of the validator set. i.e. If we have 10,000 validators, and M=1,000 then only 501 = 5% of proposers are needed to nominate any proposal.  

I think this is fine, probability of > 501 unique proposers in a 1,000 slots is high.